topic/machine-learning/llama-index/demo_mcp.py

```diff
@@ -0,0 +1,86 @@
+"""
+Use an LLM to query a database in human language via MCP.
+Example code using LlamaIndex with vanilla Open AI and Azure Open AI.
+
+https://github.com/run-llama/llama_index/tree/main/llama-index-integrations/tools/llama-index-tools-mcp
+
+## Start CrateDB MCP Server
+```
+export CRATEDB_CLUSTER_URL="http://localhost:4200/"
+cratedb-mcp serve --transport=streamable-http
+```
+
+## Usage
+```
+source env.standalone
+export OPENAI_API_KEY=sk-XJZ7pfog5Gp8Kus8D--invalid--0CJ5lyAKSefZLaV1Y9S1
+python demo_mcp.py
+```
+"""
+import asyncio
+import os
+
+from cratedb_about.instruction import Instructions
+
+from dotenv import load_dotenv
+from llama_index.core.agent.workflow import FunctionAgent
+from llama_index.llms.openai import OpenAI
+from llama_index.tools.mcp import BasicMCPClient, McpToolSpec
+
+from boot import configure_llm
+
+
+class Agent:
+
+    async def get_tools(self):
+        # Connect to the CrateDB MCP server using `streamable-http` transport.
+        mcp_url = os.getenv("CRATEDB_MCP_URL", "http://127.0.0.1:8000/mcp/")
+        mcp_client = BasicMCPClient(mcp_url)
+        mcp_tool_spec = McpToolSpec(
+            client=mcp_client,
+            # Optional: Filter the tools by name
+            # allowed_tools=["tool1", "tool2"],
+            # Optional: Include resources in the tool list
+            # include_resources=True,
+        )
+        return await mcp_tool_spec.to_tool_list_async()
+
+    async def get_agent(self):
+        return FunctionAgent(
+            name="Agent",
+            description="CrateDB text-to-SQL agent",
+            llm=OpenAI(model="gpt-4o"),
+            tools=await self.get_tools(),
+            system_prompt=Instructions.full(),
+        )
```

Comment:
Better use the full instructions including how to use the provided tools after this patch has been merged.

- https://github.com/crate/cratedb-mcp/pull/53